File: ios/Classes/POBRewardedClient.swift
import UIKit
import OpenWrapSDK
import Flutter

/// Wrapper class around `POBRewardedAd` to transfer the API calls
class POBRewardedClient: POBAdClient, POBRewardedAdDelegate {
    var rewardedAd: POBRewardedAd?
    var methodChannel: FlutterMethodChannel?

    /// Initiates a new rewarded client
    /// - Parameters:
    ///
    /// - adIndex: index to identify the client instance
    /// - pubId: Identifier of the publisher
    /// - profileId: OpenWrap profile ID
    /// - adUnitId: Ad unit id used to identify unique placement on screen
    /// - channel: `FlutterMethodChannel` for native-dart communication
    init(adIndex: Int, publisherId: String, profileId: NSNumber, adUnitId: String, channel: FlutterMethodChannel) {
        super.init(adId: adIndex)
        rewardedAd = POBRewardedAd(publisherId: publisherId, profileId: profileId, adUnitId: adUnitId)
        request = rewardedAd?.request
        methodChannel = channel
    }

    override func invokeMethod(names: [String], call: FlutterMethodCall, result: FlutterResult) {
        switch names[1] {
        case "loadAd":
            rewardedAd?.loadAd()
            result(nil)
        case "show":
            rewardedAd?.show(from: UIApplication.presentedViewController())
            result(nil)
        case "isReady":
            result(rewardedAd?.isReady)
        case "setListener":
            rewardedAd?.delegate = self
            result(nil)
        case "setSkipAlertDialogInfo":
            if let skipAlertInfo = call.arguments as? [String: Any] {
                setSkipAlert(info: skipAlertInfo)
            }
            result(nil)
        case "getBid":
            result(convertBidToMap(bid: rewardedAd?.bid()))
        case "destroy":
            // ad is removed from super class switch case
            rewardedAd = nil
            result(nil)
        default:
            super.invokeMethod(names: names, call: call, result: result)
        }
    }

    func setSkipAlert(info: [String: Any]) {
        if let title = info["title"] as? String,
           let msg = info["message"] as? String,
           let resumeTitle = info["resumeTitle"] as? String,
           let closeTitle = info["closeTitle"] as? String {
            rewardedAd?.setSkipAlertInfoWithTitle(title,
                                                  message: msg,
                                                  closeButtonTitle: closeTitle,
                                                  resumeButtonTitle: resumeTitle)
        }
    }

    // MARK: Rewarded delegate methods
    
    // Notifies the delegate that an ad has been received successfully.
    func rewardedAdDidReceive(_ rewardedAd: POBRewardedAd) {
        methodChannel?.invokeMethod("onAdReceived", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate that a user will be rewarded once the ad is completely viewed.
    func rewardedAd(_ rewardedAd: POBRewardedAd, shouldReward reward: POBReward) {
        let rewardInfo = ["currencyType": reward.currencyType, 
                          "amount": Int(truncating: reward.amount),
                          Constants.kAdIdKey: adId] as [String: Any]
        methodChannel?.invokeMethod("onReceiveReward", arguments: rewardInfo)
    }

    // Notifies the delegate of an error encountered while loading or rendering an ad.
    func rewardedAd(_ rewardedAd: POBRewardedAd, didFailToReceiveAdWithError error: Error) {
        let errMap = POBUtils.convertErrorAndAdIdToMap(error: error as NSError, adId: adId)
        methodChannel?.invokeMethod("onAdFailedToLoad", arguments: errMap)
    }
    
    // Notifies the delegate of an error encountered while showing an ad.
    func rewardedAd(_ rewardedAd: POBRewardedAd, didFailToShowAdWithError error: Error) {
        let errMap = POBUtils.convertErrorAndAdIdToMap(error: error as NSError, adId: adId)
        methodChannel?.invokeMethod("onAdFailedToShow", arguments: errMap)
    }
    
    // Notifies the delegate that the rewarded ad will be presented as a modal
    func rewardedAdWillPresent(_ rewardedAd: POBRewardedAd) {
        // no op
    }

    // Notifies the delegate that the rewarded ad has been presented as a modal
    func rewardedAdDidPresent(_ rewardedAd: POBRewardedAd) {
        methodChannel?.invokeMethod("onAdOpened", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate that the rewarded ad has been animated off the screen.
    func rewardedAdDidDismiss(_ rewardedAd: POBRewardedAd) {
        methodChannel?.invokeMethod("onAdClosed", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate of ad click
    func rewardedAdDidClick(_ rewardedAd: POBRewardedAd) {
        methodChannel?.invokeMethod("onAdClicked", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate that a user interaction will open another app (e.g. App Store), leaving the current app.
    func rewardedAdWillLeaveApplication(_ rewardedAd: POBRewardedAd) {
        methodChannel?.invokeMethod("onAppLeaving", arguments: POBUtils.adIdArgs(adId: adId))
    }
    
    // Notifies the delegate that the ad has expired
    func rewardedAdDidExpireAd(_ rewardedAd: POBRewardedAd) {
        methodChannel?.invokeMethod("onAdExpired", arguments: POBUtils.adIdArgs(adId: adId))
    }

    // Notifies the delegate that the ad impression
    func rewardedAdDidRecordImpression(_ rewardedAd: POBRewardedAd) {
        methodChannel?.invokeMethod("onAdImpression", arguments: POBUtils.adIdArgs(adId: adId))
    }
}

File: android/src/main/kotlin/com/pubmatic/sdk/flutter_openwrap_sdk/rewarded/POBRewardedClient.kt
package com.pubmatic.sdk.flutter_openwrap_sdk.rewarded

import android.content.Context
import com.pubmatic.sdk.common.POBError
import com.pubmatic.sdk.flutter_openwrap_sdk.POBAdClient
import com.pubmatic.sdk.flutter_openwrap_sdk.POBFlutterConstants
import com.pubmatic.sdk.flutter_openwrap_sdk.POBUtils
import com.pubmatic.sdk.openwrap.core.POBReward
import com.pubmatic.sdk.rewardedad.POBRewardedAd
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Wrapper around [POBRewardedAd] to transfer dart method calls to OpenWrap SDK.
 *
 * @property channel [MethodChannel] for native-dart communication
 * @constructor
 * Initializes the [POBRewardedAd] with necessary properties
 *
 * @param adId Unique Instance manager id
 * @param channel MethodChannel for triggering callback on dart side
 * @param context Application context
 * @param pubId Identifier of the publisher
 * @param profileId Profile ID of an ad tag
 * @param adUnitId Ad unit id used to identify unique placement on screen
 */
class POBRewardedClient(
  adId: Int,
  channel: MethodChannel,
  context: Context,
  pubId: String,
  profileId: Int,
  adUnitId: String
) : POBAdClient(adId, channel) {

  private val rewardedAd: POBRewardedAd? =
    POBRewardedAd.getRewardedAd(context, pubId, profileId, adUnitId)

  init {
    request = rewardedAd?.adRequest
    impression = rewardedAd?.impression
  }

  override fun methodCall(names: List<String>, call: MethodCall, result: MethodChannel.Result) {
    when (names[1]) {

      "setListener" -> {
        rewardedAd?.setListener(POBRewardedAdListenerImpl(channel, adId))
        result.success(null)
      }

      "setSkipAlertDialogInfo" -> setSkipAlertDialogInfoParams(call, result)

      "loadAd" -> {
        rewardedAd?.loadAd()
        result.success(null)
      }

      "show" -> {
        rewardedAd?.show()
        result.success(null)
      }

      "isReady" -> result.success(rewardedAd?.isReady)

      "getBid" -> result.success(convertBidToMap(rewardedAd?.bid))

      "destroy" -> {
        super.destroy()
        rewardedAd?.destroy()
        result.success(null)
      }

      else -> super.methodCall(names, call, result)
    }
  }

  private fun setSkipAlertDialogInfoParams(call: MethodCall, result: MethodChannel.Result) {
    call.argument<String>("title")?.let { title ->
      call.argument<String>("message")?.let { message ->
        call.argument<String>("resumeTitle")?.let { resumeTitle ->
          call.argument<String>("closeTitle")?.let { closeTitle ->
            rewardedAd?.setSkipAlertDialogInfo(title, message, resumeTitle, closeTitle)
            result.success(null)
            return
          }
        }
      }
    }
    result.error(
      POBFlutterConstants.OPENWRAP_PLATFORM_EXCEPTION,
      "Error while calling setSkipAlertDialogInfoParams on POBRewardedAd class.",
      "Cannot set skip alert dialog info params as one of the received values from " +
              "title, message, resumeTitle and closeTitle is null."
    )
  }

  /**
   * Listener for receiving callbacks of [POBRewardedAd] events from OpenWrapSDK.
   */
  private class POBRewardedAdListenerImpl(
    private val channel: MethodChannel,
    private val adId: Int
  ) : POBRewardedAd.POBRewardedAdListener() {

    override fun onAdReceived(rewardedAd: POBRewardedAd) {
      channel.invokeMethod("onAdReceived", POBUtils.getArgumentMap(adId))
    }

    override fun onAdFailedToLoad(rewardedAd: POBRewardedAd, error: POBError) {
      channel.invokeMethod("onAdFailedToLoad", POBUtils.getArgumentMap(adId, error))
    }

    override fun onAdFailedToShow(rewardedAd: POBRewardedAd, error: POBError) {
      channel.invokeMethod("onAdFailedToShow", POBUtils.getArgumentMap(adId, error))
    }

    override fun onAppLeaving(rewardedAd: POBRewardedAd) {
      channel.invokeMethod("onAppLeaving", POBUtils.getArgumentMap(adId))
    }

    override fun onAdOpened(rewardedAd: POBRewardedAd) {
      channel.invokeMethod("onAdOpened", POBUtils.getArgumentMap(adId))
    }

    override fun onAdClosed(rewardedAd: POBRewardedAd) {
      channel.invokeMethod("onAdClosed", POBUtils.getArgumentMap(adId))
    }

    override fun onAdClicked(rewardedAd: POBRewardedAd) {
      channel.invokeMethod("onAdClicked", POBUtils.getArgumentMap(adId))
    }

    override fun onAdExpired(rewardedAd: POBRewardedAd) {
      channel.invokeMethod("onAdExpired", POBUtils.getArgumentMap(adId))
    }

    override fun onReceiveReward(rewardedAd: POBRewardedAd, reward: POBReward) {
      val arguments: HashMap<String, Any> = POBUtils.getArgumentMap(adId) as HashMap<String, Any>
      arguments["currencyType"] = reward.currencyType
      arguments["amount"] = reward.amount
      channel.invokeMethod("onReceiveReward", arguments)
    }

    override fun onAdImpression(rewardedAd: POBRewardedAd) {
      channel.invokeMethod("onAdImpression", POBUtils.getArgumentMap(adId))
    }
  }
}